src/EditorFeatures/Text/Extensions.cs

```diff
@@ -12,7 +12,7 @@ namespace Microsoft.CodeAnalysis.Text
     public static partial class Extensions
     {
         public static SourceTextContainer AsTextContainer(this ITextBuffer buffer)
-            => TextBufferContainer.From(buffer);
+            => TextBufferContainer.From(workspace: null, buffer);
```

Comment:
This one worries me, since it's entirely likely the very first use of an ITextBuffer goes through this API (at least it is in my project system refactoring), which then means it's not going to get a workspace and never will. This may very well be causing a perf regression if the cloning stuff is necessary. Will probably have to ask for an RPS run then prior to this being merged unless we have an alternate solution, since taking an RPS regression now would cause codeflow to miss a few deadlines.

To be honest, it looks like this is all so we can call ITextBufferCloneService.Clone(ITextImage), which we could just have a static function that's Function<ITextImage, ITextBuffer> and we set that in VisualStudioWorkspace so it's OK for VS, and not set it in tests. Yes, statics are evil, but hey, at least instead of the static for PrimaryWorkspace now it's a single-use static that doesn't open pandora's box. Wishing I did that years ago now...

And that lives long enough until we pester @olegtk or somebody to give us a proper API here, because ITextImage.WithChanges() that returns an ITextImage would be both straightforward and also really useful.

Reply:
Can you review commit c7eb42d1bc224106b4e563bb598a4e575f117d86 and let me know if the approach in that change is preferable? If so I will add it to this pull request.

Reply:
➡️ Fixed in 4924d70